proxy/http2/Http2ClientSession.cc

```diff
@@ -350,13 +362,25 @@ Http2ClientSession::main_event_handler(int event, void *edata)
     break;
   }
 
-  if (!this->is_draining()) {
+  if (!this->is_draining() && this->connection_state.get_shutdown_reason() == Http2ErrorCode::HTTP2_ERROR_MAX) {
     this->connection_state.set_shutdown_state(HTTP2_SHUTDOWN_NONE);
   }
 
-  // For a case we already checked Connection header and it didn't exist
-  if (this->is_draining() && this->connection_state.get_shutdown_state() == HTTP2_SHUTDOWN_NONE) {
-    this->connection_state.set_shutdown_state(HTTP2_SHUTDOWN_NOT_INITIATED);
+  if (this->connection_state.get_shutdown_state() == HTTP2_SHUTDOWN_NONE) {
+    if (this->is_draining()) { // For a case we already checked Connection header and it didn't exist
+      Http2SsnDebug("Preparing for graceful shutdown because of draining state");
+      this->connection_state.set_shutdown_state(HTTP2_SHUTDOWN_NOT_INITIATED);
+    } else if (this->connection_state.get_stream_error_rate() >
+               Http2::stream_error_rate_threshold) { // For a case many stream errors happened
+      ip_port_text_buffer ipb;
+      const char *client_ip = ats_ip_ntop(get_client_addr(), ipb, sizeof(ipb));
```

Comment:
I kinda wish we'd put some empty lines between "code" and variable declarations, but alas, we have no coding standard dictating this. :).

Reply:
Well, I agree to the basic idea but it's not that simple I think, because some variables are only for code in the same *block* (not scope). I'd say some of them are intentionally coupled (I do this sometimes).

```diff
diff --git a/proxy/http2/Http2ClientSession.cc b/proxy/http2/Http2ClientSession.cc
index 14339ddd7..7b297c52b 100644
--- a/proxy/http2/Http2ClientSession.cc
+++ b/proxy/http2/Http2ClientSession.cc
@@ -208,9 +208,11 @@ Http2ClientSession::set_upgrade_context(HTTPHdr *h)
   unsigned char out_buf[svlen];
   if (sv && svlen > 0) {
     size_t decoded_len;
+
     ats_base64_decode(sv, svlen, out_buf, svlen, &decoded_len);
     for (size_t nbytes = 0; nbytes < decoded_len; nbytes += HTTP2_SETTINGS_PARAMETER_LEN) {
       Http2SettingsParameter param;
+
       if (!http2_parse_settings_parameter(make_iovec(out_buf + nbytes, HTTP2_SETTINGS_PARAMETER_LEN), param) ||
           !http2_settings_parameter_is_valid(param)) {
         // TODO ignore incoming invalid parameters and send suitable SETTINGS
@@ -320,6 +322,7 @@ Http2ClientSession::main_event_handler(int event, void *edata)
   case VC_EVENT_READ_COMPLETE:
   case VC_EVENT_READ_READY: {
     bool is_zombie = connection_state.get_zombie_event() != nullptr;
+
     retval         = (this->*session_handler)(event, edata);
     if (is_zombie && connection_state.get_zombie_event() != nullptr) {
       Warning("Processed read event for zombie session %" PRId64, connection_id());
@@ -329,6 +332,7 @@ Http2ClientSession::main_event_handler(int event, void *edata)
 
   case HTTP2_SESSION_EVENT_XMIT: {
     Http2Frame *frame = (Http2Frame *)edata;
+
     total_write_len += frame->size();
     write_vio->nbytes = total_write_len;
     frame->xmit(this->write_buffer);
@@ -374,6 +378,7 @@ Http2ClientSession::main_event_handler(int event, void *edata)
                Http2::stream_error_rate_threshold) { // For a case many stream errors happened
       ip_port_text_buffer ipb;
       const char *client_ip = ats_ip_ntop(get_client_addr(), ipb, sizeof(ipb));
+
       Error("HTTP/2 session error client_ip=%s session_id=%" PRId64
             " closing a connection, because its stream error rate (%f) exceeded the threshold (%f)",
             client_ip, connection_id(), this->connection_state.get_stream_error_rate(), Http2::stream_error_rate_threshold);
@@ -498,6 +503,7 @@ int
 Http2ClientSession::state_complete_frame_read(int event, void *edata)
 {
   VIO *vio = (VIO *)edata;
+
   STATE_ENTER(&Http2ClientSession::state_complete_frame_read, event);
   ink_assert(event == VC_EVENT_READ_COMPLETE || event == VC_EVENT_READ_READY);
   if (this->sm_reader->read_avail() < this->current_hdr.length) {
@@ -535,6 +541,7 @@ Http2ClientSession::state_process_frame_read(int event, VIO *vio, bool inside_fr
   while (this->sm_reader->read_avail() >= (int64_t)HTTP2_FRAME_HEADER_LEN) {
     // Return if there was an error
     Http2ErrorCode err;
+
     if (do_start_frame_read(err) < 0) {
       // send an error if specified.  Otherwise, just go away
       if (err > Http2ErrorCode::HTTP2_ERROR_NO_ERROR) {
```